File: cpp/geometry/polygon_bb4_test.py
#!/usr/bin/python3
#\file    polygon_bb4_test.py
#\brief   Test code of polygon_bb4.cpp
#\author  Akihiko Yamaguchi, info@akihikoy.net
#\version 0.1
#\date    Oct.22, 2025
import subprocess, sys, numpy as np

def run_bbox_cpp(points, exe='./polygon_bb4_test.out'):
  # Prepare stdin payload
  payload = "\n".join(
    " ".join(str(float(x)) for x in p[:3]) for p in points
  ).encode('utf-8')

  proc = subprocess.Popen([exe], stdin=subprocess.PIPE,
                          stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  out, err = proc.communicate(payload)
  if proc.returncode != 0:
    sys.stderr.write(err.decode())
    raise RuntimeError(f"{exe} failed")
  sys.stderr.write(err.decode())

  vals = [float(v) for v in out.decode().strip().split()]
  if len(vals) != 5:
    raise ValueError("Unexpected output format")
  cx, cy, w, h, ang = vals
  return (cx, cy), (w, h), ang


if __name__=='__main__':
  # Data selection
  from gen_data import *
  choice= input("Select dataset (1:Gen3d_01, 2:Gen3d_02, 3:Gen3d_11, 4:Gen3d_12, 5:Gen3d_13, 6:Gen3d_14, 7:Gen3d_14(dense=1), 8:Gen3d_15, 9:Gen3d_15(dense=1), a:Gen2d_01 [default]): ").strip()
  if choice == "1":
    points= To2d2(Gen3d_01())
  elif choice == "2":
    points= To2d2(Gen3d_02())
  elif choice == "3":
    points= To2d2(Gen3d_11())
  elif choice == "4":
    points= To2d2(Gen3d_12())
  elif choice == "5":
    points= To2d2(Gen3d_13())
  elif choice == "6":
    points= To2d2(Gen3d_14())
  elif choice == "7":
    points= To2d2(Gen3d_14(dense=1.0))
  elif choice == "8":
    points= To2d2(Gen3d_15())
  elif choice == "9":
    points= To2d2(Gen3d_15(dense=1.0))
  else:
    points= Gen2d_01()
  #print(points)


  #center,size,angle= BoundingBoxWithEllipseAxis(points)
  methods={
    '1': ('./polygon_bb2.out', 'MinAreaRect'),
    '2': ('./polygon_bb4_test.out', 'BoundingBoxWithEllipseAxis'),
    }

  for key, (exe, title) in methods.items():
    center,size,angle= run_bbox_cpp(points, exe=exe)

    rot= np.array([[np.cos(angle),-np.sin(angle)],[np.sin(angle),np.cos(angle)]])
    rect= [np.array(center)+rot.dot(p) for p in [[size[0]*0.5,size[1]*0.5],[size[0]*0.5,-size[1]*0.5],[-size[0]*0.5,-size[1]*0.5],[-size[0]*0.5,size[1]*0.5]]]

    print(f"{title}:")
    print("  Rotation angle:", angle, "rad  (", angle*(180/math.pi), "deg )")
    print("  Width:", size[0], " Height:", size[1], "  Area:", size[0]*size[1])
    print("  Center point: \n", center)
    print("  Corner points: \n", rect)

    #Plot with matplotlib
    import matplotlib.pyplot as plt

    # Convert to arrays for easy slicing
    orig= np.array(points)
    corners= np.array(rect)
    center= np.array(center)

    # Close the loops for orig and corners
    orig_closed= np.vstack([orig, orig[:1]])
    corners_closed= np.vstack([corners, corners[:1]])

    # Plot original points
    plt.plot(orig_closed[:,0], orig_closed[:,1], '--', linewidth=3.0, label='original', zorder=1)

    # Plot bounding box corners
    plt.plot(corners_closed[:,0], corners_closed[:,1], '-.', linewidth=1.2, label='bounding box', zorder=3)

    # Plot center point
    plt.plot(center[0], center[1], 'o', markersize=6, label='center', zorder=4)

    # Keep aspect ratio 1:1 like gnuplot's set size ratio -1
    ax= plt.gca()
    ax.set_aspect('equal', adjustable='box')

    plt.title(f'{title}')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.grid(True)
    plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
    plt.tight_layout()
    plt.show()

File: cpp/geometry/polygon_bb4_test.cpp
//-------------------------------------------------------------------------------------------
/*! \file    polygon_bb4_test.cpp
    \brief   Test code of polygon_bb4.cpp with the Python program.
    \author  Akihiko Yamaguchi, info@akihikoy.net
    \version 0.1
    \date    Oct.22, 2025

#Compile with Eigen solver and OpenCV:
g++ -std=c++17 -O3 -o polygon_bb4_test.out polygon_bb4_test.cpp polygon_bb4.cpp -I/usr/include/eigen3 `pkg-config --cflags --libs opencv4`

#Test:
./polygon_bb4_test.py
*/
//-------------------------------------------------------------------------------------------
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <opencv2/core.hpp>
#include <chrono>
#include "polygon_bb4.h"
//-------------------------------------------------------------------------------------------
namespace trick
{
}
//-------------------------------------------------------------------------------------------
using namespace std;
// using namespace boost;
using namespace trick;
//-------------------------------------------------------------------------------------------
// #define print(var) PrintContainer((var), #var"= ")
// #define print(var) std::cout<<#var"= "<<(var)<<std::endl
//-------------------------------------------------------------------------------------------

int main(int argc, char**argv)
{
  std::vector<cv::Vec3f> pts;
  std::string line;
  while (std::getline(std::cin, line))
  {
    if (line.empty() || line[0] == '#') continue;
    std::istringstream ss(line);
    float x, y, z = 0.0f;
    if (!(ss >> x >> y)) continue;
    ss >> z;  // optional
    pts.emplace_back(x, y, z);
  }

  if (pts.empty())
  {
    std::cerr << "No valid points received.\n";
    return 1;
  }

  auto t0 = std::chrono::high_resolution_clock::now();
  auto r = BoundingBoxWithEllipseAxis(pts, 0.0f, 1e-3f, TAngleMode::amSymmetric);
  auto t1 = std::chrono::high_resolution_clock::now();
  double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

  std::cout.setf(std::ios::fixed);
  std::cout.precision(9);
  std::cout << r.center.x << " " << r.center.y << " "
            << r.size.width << " " << r.size.height << " "
            << r.angle << std::endl;

  std::cerr << "BoundingBoxWithEllipseAxis(C++) computed with "<<ms<<" ms"<<std::endl;

  return 0;
}
//-------------------------------------------------------------------------------------------
